state-reconstruct-fetcher/src/l1_fetcher.rs

```diff
@@ -582,6 +597,94 @@ impl L1Fetcher {
         }))
     }
 
+    /// Use binary-search to find the Ethereum block on which a particular batch
+    /// was published.
+    pub async fn map_l1_batch_to_l1_block(&self, target_batch_number: U256) -> Result<U64> {
+        let event = self.contracts.v1.events_by_name("BlockCommit")?[0].clone();
+        let provider = self.provider.clone();
+
+        let mut lower_block_number = U64::from(GENESIS_BLOCK);
+        let mut upper_block_number = L1Fetcher::get_last_l1_block_number(&provider).await?;
+
+        let mut current_block_number = (upper_block_number + lower_block_number) / 2;
+        loop {
+            let mut target_is_higher = false;
+            let mut target_is_lower = false;
+
+            let filter = Filter::new()
+                .address(ZK_SYNC_ADDR.parse::<Address>().unwrap())
+                .topic0(event.signature())
+                .from_block(current_block_number)
+                .to_block(current_block_number + BLOCK_STEP);
+
+            if let Ok(logs) =
```

Comment:
so, should this loop forever in case of network errors? Then the function really should be cancellable...

Reply:
Fixed